core/trino-main/src/main/java/io/trino/operator/TableFunctionOperator.java

```diff
@@ -0,0 +1,566 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.primitives.Ints;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.trino.memory.context.LocalMemoryContext;
+import io.trino.operator.TableFunctionPartition.PassThroughColumnSpecification;
+import io.trino.spi.Page;
+import io.trino.spi.connector.SortOrder;
+import io.trino.spi.ptf.ConnectorTableFunctionHandle;
+import io.trino.spi.ptf.TableFunctionProcessor.TableFunctionProcessorProvider;
+import io.trino.spi.type.Type;
+import io.trino.sql.planner.plan.PlanNodeId;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.function.BiPredicate;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkPositionIndex;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.concat;
+import static io.trino.spi.connector.SortOrder.ASC_NULLS_LAST;
+import static java.util.Collections.nCopies;
+import static java.util.Objects.requireNonNull;
+
+public class TableFunctionOperator
+        implements Operator
+{
+    public static class TableFunctionOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+
+        // a provider of table function processor to be compiled once per partition
```

Comment:
What does "compiled" mean in this context? If this is about compiling bytecode, then it should only be done once per query / per node.

Reply:
"Compiled" is a wrong word. I changed it to "called". 

The `TableFunctionProcessorProvider` is invoked once per partition tuple (or per split in case of the source operator) to provide the `TableFunctionProcessor`. It is expected that the returned Processor has clear state. In my testing implementations, the Provider just creates a new instance of `TableFunctionProcessor` for each partition tuple / split. However, the `TableFunctionProcessorProvider` is obtained from the connector in the `LocalExecutionPlanner`. It is up to the PTF author to decide what happens on the node level, and what is the lifecycle of a `TableFunctionProcessor` it provides.